autowarefoundation/autoware.universe
# feat(operation_mode_transition_manager): support ad api

Signed-off-by: Takagi, Isamu <isamu.takagi@tier4.jp>

## Description

The operation mode transition manager supports AD API. State transition is unified as follows (local mode is same as remote).
When in vehicle control mode (disengage), the operation mode can be changed immediately without checking. When the control mode changes to autoware (engage) or the operation mode changes in autoware control mode (e.g. stop to autonomous), the operation mode will be transition mode. The transition mode exits when the conditions are satisfied.

![control-operation-mode-state-transition drawio](https://user-images.githubusercontent.com/43976882/183360825-1f914a45-d4ad-4c28-9250-b480fa305ce9.png)

## Related links

https://github.com/tier4/tier4_autoware_msgs/pull/46

## Tests performed

- mode change in vehicle control mode
    1. change to vehicle control mode if not
    2. change to autonomous mode and check if it completes without transition mode
- stop to auto (in autoware control)
    1. change to autoware control mode if not
    2. change to autonomous mode
    3. check the mode is autonomous transition, and then transition flag is false
    4. check the mode is autonomous transition, mode returns to stop if transition failed
- vehicle control to auto (in autoware control)
    1. change to vehicle control mode if not
    1. change to autonomous mode if not
    1. change to autoware control mode
    1. check the mode is autonomous transition, and then transition flag is false
    1. check the mode is autonomous transition, mode returns to vehicle control if transition failed

## Notes for reviewers

By default, the transition time is too short for testing. It can be changed with this parameter.
launch/tier4_control_launch/config/operation_mode_transition_manager/operation_mode_transition_manager.param.yaml
```yaml
  stable_check:
    duration: 0.1
```

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
